### ISSUE TITLE
vmalert missing data after upgrade 1.93 -> 1.96

### ISSUE THREAD

**Issue by sergeyshevch** [VictoriaMetrics/VictoriaMetrics#5659]
### Describe the bug

After upgrading vm stack from 1.93 to 1.96 we experienced issues on vmalert recording rules. It has gaps in data. As a workaround we specified nocache=1 on datasource url and now it works as expected

More context and recording rules spec: https://t.me/VictoriaMetrics_ru1/53712

### To Reproduce

- Deploy vm stack with next extraArgs and version v1.93

vmselect:
```
replicationFactor: "2"
search.logSlowQueryDuration: "10s"
search.maxQueryLen: "102400"
search.latencyOffset: "45s"
search.maxQueryDuration: "60s"
dedup.minScrapeInterval: 20s
vmalert.proxyURL: "http://vmalert:8080"
memory.allowedPercent: "90"
```

vmstorage:
```
dedup.minScrapeInterval: 20s
```

vmalert:
```
rule.evalDelay: 45s
```

- Create some recording rule like:

<details><summary>Details</summary>
<p>

```
apiVersion: operator.victoriametrics.com/v1beta1
kind: VMRule
metadata:
  creationTimestamp: '2023-10-02T14:00:25Z'
  generation: 3
  labels:
    argocd.argoproj.io/instance: analytics-slos
    pyrra.dev/cluster: analytics
    pyrra.dev/service: vmselect
    pyrra.dev/team: sre
  name: vmselect
  namespace: victoriametrics
spec:
  groups:
    - interval: 1m
      name: vmselect-increase
      rules:
        - expr: 'sum(increase(vm_requests_total{container="vmselect"}[1w]))'
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            slo: vmselect
            team: sre
          record: 'vm_requests:increase1w'
        - alert: SLOMetricAbsent
          annotations:
            description: >
              SLO {{ $labels.slo}} Budget in {{ $labels.namespace }} namespace,
              {{ $labels.cluster }} cluster will burn in {{ $labels.short }}/{{
              $labels.long }}.
          for: 1m
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            severity: critical
            slo: vmselect
            team: sre
        - expr: 'sum(increase(vm_request_errors_total{container="vmselect"}[1w]))'
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            slo: vmselect
            team: sre
          record: 'vm_request_errors:increase1w'
        - alert: SLOMetricAbsent
          annotations:
            description: >
              SLO {{ $labels.slo}} Budget in {{ $labels.namespace }} namespace,
              {{ $labels.cluster }} cluster will burn in {{ $labels.short }}/{{
              $labels.long }}.
          expr: 'absent(vm_request_errors_total{container="vmselect"}) == 1'
          for: 1m
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            severity: critical
            slo: vmselect
            team: sre
    - interval: 30s
      name: vmselect
      rules:
        - expr: >-
            sum(rate(vm_request_errors_total{container="vmselect"}[1m])) /
            sum(rate(vm_requests_total{container="vmselect"}[1m]))
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            slo: vmselect
            team: sre
          record: 'vm_requests:burnrate1m'
        - expr: >-
            sum(rate(vm_request_errors_total{container="vmselect"}[8m])) /
            sum(rate(vm_requests_total{container="vmselect"}[8m]))
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            slo: vmselect
            team: sre
          record: 'vm_requests:burnrate8m'
        - expr: >-
            sum(rate(vm_request_errors_total{container="vmselect"}[15m])) /
            sum(rate(vm_requests_total{container="vmselect"}[15m]))
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            slo: vmselect
            team: sre
          record: 'vm_requests:burnrate15m'
        - expr: >-
            sum(rate(vm_request_errors_total{container="vmselect"}[30m])) /
            sum(rate(vm_requests_total{container="vmselect"}[30m]))
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            slo: vmselect
            team: sre
          record: 'vm_requests:burnrate30m'
        - expr: >-
            sum(rate(vm_request_errors_total{container="vmselect"}[1h30m])) /
            sum(rate(vm_requests_total{container="vmselect"}[1h30m]))
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            slo: vmselect
            team: sre
          record: 'vm_requests:burnrate1h30m'
        - expr: >-
            sum(rate(vm_request_errors_total{container="vmselect"}[6h])) /
            sum(rate(vm_requests_total{container="vmselect"}[6h]))
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            slo: vmselect
            team: sre
          record: 'vm_requests:burnrate6h'
        - expr: >-
            sum(rate(vm_request_errors_total{container="vmselect"}[1d])) /
            sum(rate(vm_requests_total{container="vmselect"}[1d]))
          labels:
            cluster: analytics
            container: vmselect
            service: vmselect
            slo: vmselect
            team: sre
          record: 'vm_requests:burnrate1d'
        - alert: ErrorBudgetBurn
          annotations:
            description: >
              SLO {{ $labels.slo}} Budget in {{ $labels.namespace }} namespace,
              {{ $labels.cluster }} cluster will burn in {{ $labels.short }}/{{
              $labels.long }}.
          expr: >-
            vm_requests:burnrate1m{container="vmselect",slo="vmselect"} > (14 *
            (1-0.98)) and
            vm_requests:burnrate15m{container="vmselect",slo="vmselect"} > (14 *
            (1-0.98))
          for: 1m0s
          labels:
            cluster: analytics
            container: vmselect
            exhaustion: 12h
            long: 15m
            service: vmselect
            severity: critical
            short: 1m
            slo: vmselect
            team: sre
        - alert: ErrorBudgetBurn
          annotations:
            description: >
              SLO {{ $labels.slo}} Budget in {{ $labels.namespace }} namespace,
              {{ $labels.cluster }} cluster will burn in {{ $labels.short }}/{{
              $labels.long }}.
          expr: >-
            vm_requests:burnrate8m{container="vmselect",slo="vmselect"} > (7 *
            (1-0.98)) and
            vm_requests:burnrate1h30m{container="vmselect",slo="vmselect"} > (7
            * (1-0.98))
          for: 4m0s
          labels:
            cluster: analytics
            container: vmselect
            exhaustion: 1d
            long: 1h30m
            service: vmselect
            severity: critical
            short: 8m
            slo: vmselect
            team: sre
        - alert: ErrorBudgetBurn
          annotations:
            description: >
              SLO {{ $labels.slo}} Budget in {{ $labels.namespace }} namespace,
              {{ $labels.cluster }} cluster will burn in {{ $labels.short }}/{{
              $labels.long }}.
          expr: >-
            vm_requests:burnrate30m{container="vmselect",slo="vmselect"} > (2 *
            (1-0.98)) and
            vm_requests:burnrate6h{container="vmselect",slo="vmselect"} > (2 *
            (1-0.98))
          for: 15m0s
          labels:
            cluster: analytics
            container: vmselect
            exhaustion: 3d12h
            long: 6h
            service: vmselect
            severity: warning
            short: 30m
            slo: vmselect
            team: sre
        - alert: ErrorBudgetBurn
          annotations:
            description: >
              SLO {{ $labels.slo}} Budget in {{ $labels.namespace }} namespace,
              {{ $labels.cluster }} cluster will burn in {{ $labels.short }}/{{
              $labels.long }}.
          expr: >-
            vm_requests:burnrate1h30m{container="vmselect",slo="vmselect"} > (1
            * (1-0.98)) and
            vm_requests:burnrate1d{container="vmselect",slo="vmselect"} > (1 *
            (1-0.98))
          for: 45m0s
          labels:
            cluster: analytics
            container: vmselect
            exhaustion: 1w
            long: 1d
            service: vmselect
            severity: warning
            short: 1h30m
            slo: vmselect
            team: sre
    - interval: 30s
      name: vmselect-generic
      rules:
        - expr: '0.98'
          labels:
            cluster: analytics
            service: vmselect
            slo: vmselect
            team: sre
          record: pyrra_objective
        - expr: '604800'
          labels:
            cluster: analytics
            service: vmselect
            slo: vmselect
            team: sre
          record: pyrra_window
        - expr: >-
            1 -
            sum(vm_request_errors:increase1w{container="vmselect",slo="vmselect"}
            or vector(0)) /
            sum(vm_requests:increase1w{container="vmselect",slo="vmselect"})
          labels:
            cluster: analytics
            service: vmselect
            slo: vmselect
            team: sre
          record: pyrra_availability
        - expr: 'sum(vm_requests_total{container="vmselect"})'
          labels:
            cluster: analytics
            service: vmselect
            slo: vmselect
            team: sre
          record: pyrra_requests_total
        - expr: 'sum(vm_request_errors_total{container="vmselect"} or vector(0))'
          labels:
            cluster: analytics
            service: vmselect
            slo: vmselect
            team: sre
          record: pyrra_errors_total
```

</p>
</details> 

- Upgrade cluster to 1.96
- There should be gaps in recording rule

### Version

/ # ./vmalert-prod --version
vmalert-20231212-224707-tags-v1.96.0-0-g304fe0565

### Logs

No

### Screenshots

![telegram-cloud-photo-size-2-5291920048711130210-y](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/15252187/cae9d381-f1d1-46d9-9a94-3425c719b64d)
![telegram-cloud-photo-size-2-5291920048711130211-y](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/15252187/8b131063-a733-40a0-bc89-5c00d04ab636)
![telegram-cloud-photo-size-2-5291920048711130212-y](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/15252187/6b5598ac-30e0-4ad0-bfef-e3ca16874126)

![telegram-cloud-photo-size-2-5294171848524814505-y](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/15252187/83095470-55b9-4e2c-bcb0-1786fd59aae5)
![telegram-cloud-photo-size-2-5294171848524814529-y](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/15252187/572a1ae2-63ea-45da-ac15-33fbc4ff3b7b)

**Comment by valyala** [comment]
@sergeyshevch , could you verify whether this issue is fixed in [v1.97.0](https://docs.victoriametrics.com/CHANGELOG.html)?

**Comment by sergeyshevch** [comment]
@valyala Thanks! Now all works as expected
